### PR TITLE
Initial socketless local mode

### PR DESCRIPTION
**Pull request by danielsdeleo**
Connect to chef zero directly by converting requests into rack format and converting responses from rack to Net::HTTP.

### CODE REVIEW

File: lib/chef/rest.rb [hunk 1]
```diff
@@ -188,9 +370,25 @@ def follow_redirect
...
+        URI.parse(path)
+      else
+        super
+      end
+    end
+
     def http_client(base_url=nil)
       base_url ||= url
-      BasicClient.new(base_url, :ssl_policy => Chef::HTTP::APISSLPolicy)
+      pp url_class: base_url.class, value: base_url
```
Comment (ranjib) [inline lib/chef/rest.rb]: debug??

**Comment by ranjib** [comment]
i'll do couple of integration tests over the weekend. LGTM kitchen failures will need fixing.
:+1:

File: lib/chef/rest.rb (hunk 1 above)
Comment (danielsdeleo) [inline lib/chef/rest.rb]: Yeah, this extremely rough, I only put it up here to let people vet the design. I'm certain there are a bunch of failing integration tests (which travis might not run so I'm not sure if it'll pass travis, but it's broken anyway).

**Comment by danielsdeleo** [comment]
@ranjib awesome, this actually seems pretty solid. After chatting with @jkeiser I've decided that disabling the socket should be opt-in until Chef 13, so I'm gonna add a `--[no-]listen` flag in a future commit. For now, the default will be listen, in Chef 13 it will be "no listen". In any case, all local mode communication is socketless regardless of whether the socket is enabled.

**Comment by danielsdeleo** [comment]
Oops, didn't mean to cancel :person_frowning:

**Comment by lamont-granquist** [comment]
:+1:

**Comment by jkeiser** [comment]
I love it!  As we chatted about, the one issue we are going to have is that chef-provisioning is expecting a listen-port so it can tunnel remote clients back to your local mode chef-zero server.  The ideal thing here would be that in the short term:

1. Local mode by default starts a localhost-listening server like normal
2. chef-client in local mode communicates to that server using the bogus requests (speed, less complexity)
3. --no-listen turns off the port listening behavior

If the chef-zero server can accept both TCP and non-TCP requests, that will set us up well for a Chef 13 no-listen-by-default future: chef-provisioning can then flip TCP on, on demand, if it needs to tunnel.

**Comment by danielsdeleo** [comment]
@jkeiser yep, as of the last commit I pushed (along with https://github.com/chef/chef-zero/pull/121), the only behavior that occurs is that chef-zero binds to a port, but chef-client uses the sockets `chefzero://localhost` URL internally. So I just need to add the `--no-listen` flag and some unit tests and it should be good to go I think.

File: lib/chef/rest.rb [hunk 2]
```diff
@@ -39,6 +39,152 @@ class REST < HTTP; end
 require 'chef/http/remote_request_id'
 
 class Chef
+
+  class SocketlessChefZeroClient
+
+    module Response
+
+      def read_body(dest = nil, &block)
```
Comment (danielsdeleo) [inline lib/chef/rest.rb]: I should document why this is necessary. tl;dr, we set `@read` to true, but `Net::HTTP` won't let you do a streaming style body handler (with a block) after reading the body the first time.

File: lib/chef/rest.rb [hunk 3]
```diff
@@ -39,6 +39,152 @@ class REST < HTTP; end
...
+
+    def host
+      @url.hostname
+    end
+
+    def port
+      @url.port
+    end
+
+    # request, response = client.request(method, url, body, headers) {|r| r.read_body }
```
Comment (danielsdeleo) [inline lib/chef/rest.rb]: Leftover copypasta.

File: lib/chef/rest.rb [hunk 4]
```diff
@@ -39,6 +39,152 @@ class REST < HTTP; end
...
+
+      net_http_response = to_net_http(res[0], res[1], res[2])
+
+      yield net_http_response if block_given?
+
+      [self, net_http_response]
+    end
+
+    # TODO: this is copied verbatim from the fakeweb project, MIT licensed
+    # Add credits where appropriate
```
Comment (danielsdeleo) [inline lib/chef/rest.rb]: Reminder to myself to do this. https://github.com/chrisk/fakeweb

File: lib/chef/rest.rb [hunk 5]
```diff
@@ -39,6 +39,152 @@ class REST < HTTP; end
...
+
+    def to_net_http(code, headers, chunked_body)
+      body = chunked_body.join('')
+      msg = STATUS_MESSAGE[code] or raise "Cannot determine HTTP status message for code #{code}"
```
Comment (danielsdeleo) [inline lib/chef/rest.rb]: This case should be rare, but when it happens, it'll be helpful to have the rest of the response in the exception message so you can work out what the request was.

**Comment by danielsdeleo** [comment]
Travis/Kitchen failure is `ERROR:  Could not find a valid gem 'chef-zero' (~> 4.0) in any repository` which should get fixed when we merge and ship the Chef Zero PR.

File: Gemfile [hunk 6]
```diff
@@ -3,6 +3,8 @@ gemspec :name => "chef"
 
 gem "activesupport", "< 4.0.0", :group => :compat_testing, :platform => "ruby"
 
+gem "chef-zero", github: "chef/chef-zero", branch: "socketless"
```
Comment (danielsdeleo) [inline Gemfile]: Reminder to myself to ensure this is reverted

File: lib/chef/http.rb [hunk 7]
```diff
@@ -196,14 +196,18 @@ def streaming_request(path, headers={}, &block)
 
     def http_client(base_url=nil)
       base_url ||= url
-      BasicClient.new(base_url)
+      if chef_zero_uri?(base_url)
```
Comment (danielsdeleo) [inline lib/chef/http.rb]: Need to add tests for this branch

File: lib/chef/rest.rb [hunk 8]
```diff
@@ -57,6 +58,9 @@ class REST < HTTP
     # http://localhost:4000, a call to +get_rest+ with 'nodes' will make an
     # HTTP GET request to http://localhost:4000/nodes
     def initialize(url, client_name=Chef::Config[:node_name], signing_key_filename=Chef::Config[:client_key], options={})
+
+      signing_key_filename = nil if chef_zero_uri?(url)
```
Comment (danielsdeleo) [inline lib/chef/rest.rb]: This needs a test.

Comment (danielsdeleo) [inline lib/chef/rest.rb]: done

File: lib/chef/http.rb (hunk 7 above)
Comment (danielsdeleo) [inline lib/chef/http.rb]: done.

**Comment by danielsdeleo** [comment]
@chef/client-core This is ready for review. The kitchen tests won't pass until we merge and ship the chef-zero patch, though.

File: lib/chef/rest.rb [hunk 9]
```diff
@@ -188,11 +192,6 @@ def follow_redirect
 
     public :create_url
 
-    def http_client(base_url=nil)
-      base_url ||= url
-      BasicClient.new(base_url, :ssl_policy => Chef::HTTP::APISSLPolicy)
-    end
-
```
Comment (btm) [inline lib/chef/rest.rb]: cruft?

**Comment by btm** [comment]
:+1:

File: lib/chef/rest.rb (hunk 9 above)
Comment (danielsdeleo) [inline lib/chef/rest.rb]: Yeah, `Chef::REST` inherits from `Chef::HTTP`, which has the exact same code as was here.

**Pull request merged**